File: Data Structures & Algorithms/reverse-a-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head: return head

        arm_node = head
        new_head = head
        while arm_node.next:
            
            # reference node to rotate to head
            rotate_node = arm_node.next

            if arm_node.next.next:
                # link arm to rotating node's next
                arm_node.next = rotate_node.next
                
                # move rotating node to head
                rotate_node.next = new_head
                new_head = rotate_node
            
            else:
                # move rotating node to head
                rotate_node.next = new_head
                new_head = rotate_node

                # end while loop
                arm_node.next = None
        return new_head

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        if (head == NULL) return NULL;

        ListNode* tail = head;
        ListNode* n_front = head;
        int count = n;

        while (count-- > 0 && n_front != NULL) 
            n_front = n_front->next;
        
        // Then move both pointers at the same pace until the end of the 
        // linked list path
        while (n_front != NULL && n_front->next != NULL){
            tail = tail->next;
            n_front = n_front->next;
        }

        // at this point, tail pointer is the node before the node for removal.
        if (tail == head && tail->next == NULL) 
            head = head->next;
        else if (n_front == NULL) {
            // check the distance from tail
            ListNode* curr_node = tail;
          
            int steps = 0;
            while(curr_node->next != NULL){
                curr_node = curr_node->next;
                steps++;
            }
          
            if (n-steps == 1)
                head = head->next;
        }
        else if (tail->next != NULL){
            n_front = tail->next;
            tail->next = n_front->next;
        } 
        return head;
    }
};
